src/lib.rs

```diff
@@ -1,173 +1,73 @@
-    iVert: c_int,
-) {
-    unsafe {
-        let x = (*pContext).m_pUserData as *const Closures;
-        let slice = ((*x).tex_coord)(iFace as usize, iVert as usize);
-        let src = slice.as_ptr() as *const c_float;
-        ptr::copy_nonoverlapping::<c_float>(src, fvTexcOut, 2);
-    }
+fn get_position<I: Geometry>(geometry: &mut I, index: usize) -> Vector3<f32> {
+    let (face, vert) = index_to_face_vert(index);
+    geometry.get_position(face, vert).coords
 }
 
-/// Returns the tangent and its sign to the application.
-extern "C" fn set_tspace_basic(
-    pContext: *mut ffi::SMikkTSpaceContext,
-    fvTangent: *const c_float,
-    fSign: c_float,
-    iFace: c_int,
-    iVert: c_int,
-) {
-    unsafe {
-        let x = (*pContext).m_pUserData as *mut Closures;
-        let mut tangent: [f32; 4] = mem::uninitialized();
-        let dst: *mut c_float = tangent.as_mut_ptr();
-        ptr::copy_nonoverlapping::<c_float>(fvTangent, dst, 3);
-        tangent[3] = fSign;
-        ((*x).set_tangent)(iFace as usize, iVert as usize, tangent);
-    }
+fn get_tex_coord<I: Geometry>(geometry: &mut I, index: usize) -> Vector3<f32> {
+    let (face, vert) = index_to_face_vert(index);
+    geometry.get_tex_coord(face, vert).coords.insert_row(2, 1.0)
 }
 
-/// Returns tangent space results to the application.
-extern "C" fn set_tspace(
-    pContext: *mut ffi::SMikkTSpaceContext,
-    fvTangent: *const c_float,
-    _fvBiTangent: *const c_float,
-    _fMagS: c_float,
-    _fMagT: c_float,
-    bIsOrientationPreserving: ffi::tbool,
-    iFace: c_int,
-    iVert: c_int,
-) {
-    let fSign = if bIsOrientationPreserving != 0 { 1.0 } else { -1.0 };
-    set_tspace_basic(pContext, fvTangent, fSign, iFace, iVert);
+fn get_normal<I: Geometry>(geometry: &mut I, index: usize) -> Vector3<f32> {
+    let (face, vert) = index_to_face_vert(index);
+    geometry.get_normal(face, vert)
 }
 
-impl<'a> Closures<'a> {
-    /// Generates tangents.
-    pub fn generate(mut self) -> bool {
-        let ctx = ffi::SMikkTSpaceContext {
-            m_pInterface: &INTERFACE,
-            m_pUserData: &mut self as *mut Closures as *mut c_void,
-        };
-        unsafe {
-            ffi::genTangSpaceDefault(&ctx) == ffi::TTRUE
-        }
-    }
+fn index_to_face_vert(index: usize) -> (usize, usize) {
+    (index >> 2, index & 0x3)
 }
 
-/// Generates tangents.
-pub fn generate_tangents<'a>(
-    vertices_per_face: &'a Fn() -> usize,
-    face_count: &'a Fn() -> usize,
-    position: &'a Fn(usize, usize) -> &'a [f32; 3],
-    normal: &'a Fn(usize, usize) -> &'a [f32; 3],
-    tex_coord: &'a Fn(usize, usize) -> &'a [f32; 2],
-    set_tangent: &'a mut FnMut(usize, usize, [f32; 4]),
-) -> bool {
-    let closures = Closures {
-        vertices_per_face,
-        face_count,
-        position,
-        normal,
-        tex_coord,
-        set_tangent,
-    };
-    closures.generate()
+fn face_vert_to_index(face: usize, vert: usize) -> usize {
+    face << 2 | vert & 0x3
```

Comment:
I'm not following what's going on with these bitwise operations.

Reply:
I don't either honestly, this was directly copied from the original code.

Reply:
Looking into it more, it seems to be creating an index by simply bitshifting the face and vert values until they don't overlap and creating one value out of it, I'm thinking this should probably be changed back to a u32 again.